libgalois/src/analytics/jaccard/jaccard.cpp

```diff
@@ -0,0 +1,250 @@
+    for (const auto& e : graph_.edges(n2)) {
+      auto neighbor = graph_.GetEdgeDest(e);
+      if (base_neighbors.count(*neighbor) > 0)
+        intersection_size++;
+    }
+    return intersection_size;
+  }
+};
+
+template <typename IntersectAlgorithm>
+galois::Result<void>
+JaccardImpl(
+    galois::graphs::PropertyGraph<std::tuple<JaccardSimilarity>, std::tuple<>>&
+        graph,
+    size_t compare_node, JaccardPlan /*plan*/) {
+  if (compare_node >= graph.size()) {
+    return galois::ErrorCode::InvalidArgument;
+  }
+
+  auto it = graph.begin();
+  std::advance(it, compare_node);
+  Graph::Node base = *it;
+
+  uint32_t base_size = graph.edge_end(base) - graph.edge_begin(base);
+
+  IntersectAlgorithm intersect_with_base{graph, base};
+
+  // Compute the similarity for each node
+  galois::do_all(
+      galois::iterate(graph),
+      [&](const GNode& n2) {
+        double& n2_data = graph.GetData<JaccardSimilarity>(n2);
+        uint32_t n2_size = graph.edge_end(n2) - graph.edge_begin(n2);
+        // Count the number of neighbors of n2 and the number that are shared
+        // with base
+        uint32_t intersection_size = intersect_with_base(n2);
+        // Compute the similarity
+        uint32_t union_size = base_size + n2_size - intersection_size;
+        double similarity =
+            union_size > 0 ? (double)intersection_size / union_size : 1;
+        // Store the similarity back into the graph.
+        n2_data = similarity;
+      },
+      galois::loopname("Jaccard"));
+
+  return galois::ResultSuccess();
+}
+
+}  // namespace
+
+galois::Result<void>
+galois::analytics::Jaccard(
+    graphs::PropertyFileGraph* pfg, size_t compare_node,
+    const std::string& output_property_name, JaccardPlan plan) {
+  if (auto result =
+          ConstructNodeProperties<NodeData>(pfg, {output_property_name});
+      !result) {
+    return result.error();
+  }
+
+  auto pg_result = Graph::Make(pfg, {output_property_name}, {});
+  if (!pg_result) {
+    return pg_result.error();
+  }
+
+  galois::Result<void> r = galois::ResultSuccess();
+  switch (plan.edge_sorting()) {
+  case JaccardPlan::kUnknown:
+    // TODO(amp): It would be possible to start with the sorted case and then
+    //  fail to the unsorted case if unsorted nodes are detected.
+  case JaccardPlan::kUnsorted:
+    r = JaccardImpl<IntersectWithUnsortedEdgeList>(
+        pg_result.value(), compare_node, plan);
+    break;
+  case JaccardPlan::kSorted:
+    r = JaccardImpl<IntersectWithSortedEdgeList>(
+        pg_result.value(), compare_node, plan);
+    break;
+  }
+
+  if (!r) {
+    // Undo property creation.
+    if (auto r1 = pfg->RemoveNodeProperty(output_property_name); !r1) {
+      GALOIS_LOG_WARN("Error occurred during error handling: {}", r1.error());
+    }
+  }
+
+  return r;
+}
+
+constexpr static const double EPSILON = 1e-6;
+
+galois::Result<void>
+galois::analytics::JaccardAssertValid(
+    galois::graphs::PropertyFileGraph* pfg, size_t compare_node,
```

Comment:
`uint32_t` for compare_node?

Reply:
GNode doesn't really exist any more. size_t is wrong though. I'll fix it.